velox/functions/prestosql/ArraySum.cpp

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Set.h>
+
+#include "velox/expression/EvalCtx.h"
+#include "velox/expression/Expr.h"
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/prestosql/CheckedArithmetic.h"
+
+namespace facebook::velox::functions {
+namespace {
+///
+/// Implements the array_sum function.
+/// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+template <typename TInput, typename TOutput>
+class ArraySumFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Prepare result vector for writing
+    BaseVector::ensureWritable(rows, outputType, context->pool(), result);
+    auto resultValues = (*result)->template asFlatVector<TOutput>();
+
+    // Acquire the array elements vector.
+    auto arrayVector = args[0]->as<ArrayVector>();
+    VELOX_CHECK(arrayVector);
+    auto elementsVector = arrayVector->elements();
+
+    if (elementsVector->encoding() == VectorEncoding::Simple::FLAT) {
+      const TInput* __restrict rawElements =
+          elementsVector->as<FlatVector<TInput>>()->rawValues();
+      const uint64_t* __restrict rawNulls = elementsVector->rawNulls();
+
+      if (elementsVector->mayHaveNulls()) {
+        rows.template applyToSelected([&](vector_size_t row) {
+          int start = arrayVector->offsetAt(row);
+          int end = start + arrayVector->sizeAt(row);
```

Comment:
does it make sense to avoid code duplication and create a common function templated by the mayeHaveNull
like the one below? same for the other path.
```
template <bool mayHaveNulls>
void applyFlat(..) {
  auto start = arrayVector->offsetAt(row);
  auto end = start + arrayVector->sizeAt(row);
  TOutput sum = 0;
  for (; start < end; start++) {
    if constexpr (mayHaveNulls) {
       if(bits::isBitNull(rawNulls, start)){
         continue;
       }              
    } else {
      if constexpr (std::is_same<TOutput, int64_t>::value) {
        sum = checkedPlus<TOutput>(sum, rawElements[start]);
      } else {
        sum += rawElements[start];
      }
    }
  }
  resultValues->set(row, sum);
}
```

Reply:
Done.